File: A* Algorithm/AA.py
class Node:
  def __init__(self, position, parent=None):
    self.position = position
    self.parent = parent
    self.g = 0  # Cost from start to current node
    self.h = 0  # Heuristic estimate of cost to goal
    self.f = 0  # Total cost (g + h)

def manhattan_distance(a, b):
  # Heuristic function using Manhattan distance
  return abs(a[0] - b[0]) + abs(a[1] - b[1])

def a_star(grid, start, goal):
  rows, cols = len(grid), len(grid[0])
  open_set = []
  closed_set = set()

  start_node = Node(start)
  open_set.append(start_node)

  while open_set:
    current = min(open_set, key=lambda node: node.f)
    open_set.remove(current)
    closed_set.add(current.position)

    if current.position == goal:
      path = []
      while current:
        path.append(current.position)
        current = current.parent
      return path[::-1]  # Reverse path for start to goal

    for delta_row, delta_col in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
      new_position = (current.position[0] + delta_row, current.position[1] + delta_col)
      if 0 <= new_position[0] < rows and 0 <= new_position[1] < cols and grid[new_position[0]][new_position[1]] != 1:  # Check for walls
        if new_position in closed_set:
          continue

        parent = current
        g = current.g + 1  # Movement cost (adjust if needed)
        h = manhattan_distance(new_position, goal)
        f = g + h

        new_node = Node(new_position, parent)
        new_node.g = g
        new_node.h = h
        new_node.f = f

        if new_node not in [node for node in open_set if node.position == new_position and node.f >= f]:
          open_set.append(new_node)

  return None  # No path found

# Example usage
grid = [
  [0, 0, 1, 0],
  [0, 0, 1, 0],
  [0, 0, 0, 0],
  [0, 1, 0, 0],
]

start = (0, 0)
goal = (3, 3)

path = a_star(grid, start, goal)

if path:
  print("Path found:", path)
else:
  print("No path found")

File: A* Algorithm/AA.rs
use std::collections::HashMap;
use std::cmp::Ord;

#[derive(Clone, Eq, PartialEq, PartialOrd)]
struct Node {
    position: (i32, i32), // Replace with your node type
    cost: u32,
    heuristic: u32,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        (self.cost + self.heuristic).cmp(&(other.cost + other.heuristic))
    }
}

fn astar(
    start: &Node,
    successors: fn(&(i32, i32)) -> Vec<(i32, i32, u32)>, // Replace with your successor function
    heuristic: fn(&(i32, i32)) -> u32, // Replace with your heuristic function
    is_goal: fn(&(i32, i32)) -> bool,
) -> Option<Vec<(i32, i32)>> {
    let mut open_set: HashMap<(i32, i32), Node> = HashMap::new();
    let mut closed_set: HashMap<(i32, i32), u32> = HashMap::new();

    open_set.insert(start.position.clone(), Node {
        position: start.position.clone(),
        cost: 0,
        heuristic: heuristic(&start.position),
    });

    while !open_set.is_empty() {
        let current = open_set.iter().min().unwrap().1.clone();
        open_set.remove(&current.position);
        closed_set.insert(current.position, current.cost);

        if is_goal(&current.position) {
            let mut path = vec![current.position];
            let mut parent = current;
            while closed_set.contains_key(&parent.position) {
                let parent_pos = *closed_set.get(&parent.position).unwrap();
                path.push(parent_pos);
                parent = open_set.get(&parent_pos).unwrap().clone();
            }
            path.reverse();
            return Some(path);
        }

        for (next_pos, cost, _) in successors(&current.position) {
            if closed_set.contains_key(&next_pos) {
                continue;
            }
            let tentative_cost = current.cost + cost;
            if !open_set.contains_key(&next_pos) || tentative_cost < open_set[&next_pos].cost {
                open_set.insert(
                    next_pos,
                    Node {
                        position: next_pos,
                        cost: tentative_cost,
                        heuristic: heuristic(&next_pos),
                    },
                );
            }
        }
    }

    None
}

fn main() {
    // Define your specific data structures for nodes, neighbors, and heuristic
    let start_position = (0, 0); // Replace with your starting node position
    let goal_position = (5, 5); // Replace with your goal node position
  
    // Define functions for successors and heuristic based on your graph structure
    fn successors((x, y): &(i32, i32)) -> Vec<(i32, i32, u32)> {
      // Implement logic to find valid neighbors and their costs here
      // ...
    }
  
    fn heuristic((x, y): &(i32, i32)) -> u32 {
      // Implement logic to estimate remaining cost to goal here (e.g., Manhattan distance)
      // ...
    }
  
    fn is_goal((x, y): &(i32, i32)) -> bool {
      x == &goal_position.0 && y == &goal_position.1
    }
  
    // Call the A* function with defined arguments
    let path = astar(
        &Node { position: start_position, cost: 0, heuristic: heuristic(&start_position) },
        successors,
        heuristic,
        is_goal,
    );
  
    if let Some(path) = path {
        println!("Path found: {:?}", path);
    } else {
        println!("No path found to goal.");
    }
  }
  